File: q09/q09_simple.js
function print_array(array) {
  var ary_str = "";
  for (var i = 0; i < array.length; i++) {
    for (var j = 0; j < array[i].length; j++) {
      var space = 7;
      var count = String(array[i][j]).length;
      var s = "";
      if (String(array[i][j]).length < space) {
        space -= count;
        for (var k = 0; k < space; k++) {
          s += " "
        }
      }
      ary_str += s + ary[i][j] + " ";
    }
    ary_str += "\n";
  }
  console.log(ary_str);
}

var vertical = 3;    // 縦の数
var horizontal = 4;  // 横の数
vertical += 1;       // 配列で管理する上で都合がいい
horizontal += 1;     // 配列で管理する上で都合がいい
var ary = new Array(vertical);
for (var i = 0; i < vertical; i++) {
  ary[i] = new Array(horizontal);
  for (var j = 0; j < horizontal; j++) {
    ary[i][j] = 0;
  }
}

ary[0][0] = 1;
for (var i = 0; i < vertical; i++) {
  for( var j = 0; j < horizontal; j++) {
    if (j > 0) {
      ary[i][j] += ary[i][j - 1];
    }
    if (i > 0) {
      ary[i][j] += ary[i - 1][j];
    }
    console.log("i = " + i + " j = " + j);
    print_array(ary);
  }
}

console.log(ary[vertical - 2][horizontal - 1] + ary[vertical - 1][horizontal - 2]);

File: q09/q09_simple.rb
vertical, horizontal = 3, 4
vertical, horizontal = vertical + 1, horizontal + 1

ary = Array.new(vertical * horizontal){0}
ary[0] = 1
horizontal.times do |h|
  vertical.times do |v|
    target = v + vertical * h
    # targetのマスの1つ左のマス
    pre_pos_left = v - 1 + vertical * h
    # targetのマスの1つ下のマス
    pre_pos_bottom = v + vertical * (h -1)
    # 上記のポジションのマスを加算することで，何回通過するかが分かる
    ary[target] += ary[pre_pos] if v > 0
    ary[target] += ary[pre_pos_bottom] if h > 0
  end
end

for i in 0...ary.length
  puts "ary[#{i}] = #{ary[i]}"
end

puts ary[-2] + ary[-vertical - 1]

# ary[0] = 1
# g = 0, b = 0 do nothing
# g = 0, b = 1
# ary[1] += ary[0]
# g = 0, b = 2
# ary[2] += ary[1]
# ...
# g = 1, b = 0
# ary[21] *= ary[0]
# g = 1, b = 1 through
# g = 1, b = 2
# ary[23] += ary[22]
# ary[23] += ary[2]
